ddtrace/debugging/_exception/replay.py

```diff
@@ -47,13 +48,70 @@
 FRAME_FILE_TAG = "_dd.debug.error.%d.file"
 FRAME_LINE_TAG = "_dd.debug.error.%d.line"
 
+EXCEPTION_IDENT_LIMIT = 3600.0  # 1 hour
+EXCEPTION_IDENT_LIMITER: t.Dict[int, HourGlass] = {}
+
+
+ExceptionChain = t.Deque[t.Tuple[BaseException, t.Optional[TracebackType]]]
+
+
+def exception_ident(exc: BaseException, tb: t.Optional[TracebackType]) -> int:
+    """Compute the identity of an exception.
+
+    We use the exception type and the traceback to generate a unique identifier
+    that we can use to identify the exception. This can be used to rate limit
+    the number of times we capture information of the same exception.
+    """
+    h = 0
+    _tb = tb
+    while _tb is not None:
+        frame = _tb.tb_frame
+        h = (h << 1) ^ (id(frame.f_code) << 4 | frame.f_lasti)
+        h &= 0xFFFFFFFFFFFFFFFF
+        _tb = _tb.tb_next
+    return (id(type(exc)) << 64) | h
+
+
+def exception_chain_ident(chain: ExceptionChain) -> int:
+    h = 0
+    for exc, tb in chain:
+        h = (h << 1) ^ exception_ident(exc, tb)
+        h &= 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
+    return h
+
+
+def limit_exception(chain: ExceptionChain) -> bool:
```

Comment:
Isn't it confusing there are now two ways to rate-limit exception (can_capture & limit_exception)?

Reply:
I agree this can be confusing. The reason why I haven't unified the two is that I don't want to unwind the exception chain if the global rate limiter, or the data on the trace that we might have put before, tells us that we don't want to capture. This way we can prevent doing some throw-away work.

Reply:
We've discussed offline about ways to tackle exceptions fairness to possibly mitigate this subtlety